src/i18n.ts

```diff
@@ -0,0 +1,248 @@
+  FluentBundleOptions,
+  FluentValue,
+  LocaleId,
+  Middleware,
+  NextFunction,
+  Scope,
+  TranslationContext,
+} from "./deps.ts";
+
+import { exists, readLocalesDir } from "./utils.ts";
+
+import type {
+  I18nConfig,
+  I18nContextFlavor,
+  TranslateFunction,
+} from "./types.ts";
+
+class FluentContext extends FluentType<string> {
+  constructor(value: string, private defaultValue = "") {
+    super(value);
+  }
+
+  toString(scope: Scope): string {
+    let ctx = JSON.parse(scope.args?.ctx as string);
+    const keys = this.value.split(".");
+    for (const key of keys) {
+      ctx = ctx[key];
+      if (ctx === undefined) break;
+    }
+    return ctx ?? this.defaultValue;
+  }
+}
+
+/**
+ * A custom function for accessing `Context` properties directly from
+ * Fluent translation files.
+ */
+export function CTX(args: FluentValue[]) {
+  if (args[1] !== undefined && typeof args[1] !== "string") {
+    throw new TypeError("CTX: Only string type is allowed for defaultValue.");
+  }
+
+  const key = args[0];
+  const defaultValue = typeof args[1] === "string" ? args[1] : undefined;
+
+  if (typeof key === "string") {
+    return new FluentContext(key, defaultValue);
+  }
+
+  throw new TypeError(
+    "CTX: Invalid argument type. Both key and defaultValue should be the type of string.",
+  );
+}
+
+export class I18n<C extends Context = Context> {
+  private config: I18nConfig<C>;
+  readonly fluent: Fluent;
+  readonly locales = new Array<string>();
+
+  constructor(config: Partial<I18nConfig>) {
+    this.config = {
+      defaultLocale: "en",
+      fluentBundleOptions: { functions: {} },
+      ...config,
+    };
+
+    if (
+      this.config.fluentBundleOptions &&
+      this.config.fluentBundleOptions.functions
+    ) {
+      this.config.fluentBundleOptions.functions["CTX"] = CTX;
+    } else {
+      this.config.fluentBundleOptions = {
+        ...this.config.fluentBundleOptions,
+        functions: { CTX },
+      };
+    }
+
+    this.fluent = new Fluent(this.config.fluentOptions);
+  }
+
+  /**
+   * Load and register translations from a directory to the Fluent instance.
+   * @param directory Path to the directory to look for the translation files.
+   * Translation files in that directory should end with `.ftl` extension.
+   */
+  async loadLocalesDir(directory: string): Promise<void> {
+    if (!exists(directory)) {
+      throw new Error(`Locales directory '${directory}' not found`);
+    }
+
+    for (const file of readLocalesDir(directory)) {
+      const path = resolve(directory, file);
+      const locale = file.substring(0, file.lastIndexOf("."));
+
+      await this.loadLocale(locale, {
+        filePath: path,
+        bundleOptions: this.config.fluentBundleOptions,
+      });
+    }
+  }
+
+  /**
+   * Loads and registers a translation to the Fluent instance.
+   * @param locale Locale ID
+   * @param options Options to specify the source and behavior of the translation.
+   */
+  async loadLocale(
```

Comment:
@dcdunkan

Reply:
@KnorpelSenf In the file path? No.

Reply:
Maybe we can open an issue about this. Not 1.0 stuff, but I think people will ask for this eventually anyway.

Reply:
Let's leave this unresolved until the rewrite and released and we create an issue.

Reply:
But I don't understand why we should support glob patterns here. Can you explain?

Reply:
In my experience, wherever you support file paths to multiple files, it's only a matter of time until someone asks for that. But until it's actually needed, it would be enough to open an issue and communicate that we have thought about it, without actually investing time in the implementation. Then people can comment that they need it, which is a lower threshold than offering a plugin that just doesn't offer it.

Reply:
Maybe I'm wrong, but I don't think we actually support file paths to multiple files here 🤔. We're only expecting a single file path to be specified, because we're only allowing the user to load a single locale using this method.

```ts
await i18n.loadLocale("en", { filePath: "another-locales/en.ftl" });
```

Reply:
Oh I somehow assumed we could pass directories. My bad. Then glob patterns don't make sense.

Reply:
Resolving this.